### ISSUE TITLE
Evaluate Zuul for build.next

### ISSUE THREAD

**Issue by rohanKanojia** [openshiftio/openshift.io#3604]
Evaluate Zuul CI to check if it can be the next build service we are looking for: https://zuul-ci.org/
#3570

**Comment by jaseemabid** [comment]
Leaving some useful links. 

https://techcrunch.com/2018/05/21/openstack-spins-out-its-zuul-open-source-ci-cd-platform
https://www.businesswire.com/news/home/20180521005376/en/Open-Source-CICD-Platform-Zuul-Releases-V3

**Comment by joshuawilson** [comment]
Please add labels.

**Comment by rohanKanojia** [comment]
## Basic Architecture:
Zuul is a program that drives continuous integration, delivery and deployment systems with a focus on 
project gating and interrelated projects.

### Zuul Components:
Zuul provides the following components:
1.) **Zuul Scheduler**: The main Zuul process. Handles receiving events, executing jobs, collecting
                                results and posting reports. Coordinates the work of other componenets.
                    
2.) **Zuul merger**   : Scale out component that performs git merge operations. Zuul performs a 
                              large number of git operations in the course of its work. Adding merger
                              process can help speed Zuul's processing.

3.) **Zuul executor** : Scale out component for executing jobs. At least one of these is required
                              Depending on system configuration, you can expect a single executor to 
                              handle up to about 100 simultaneous jobs. Can handle the functions of a 
                              merger if dedicated mergers are not provided.

4.) **Zuul web**      : A web server that currently provides websocket access to live streaming
                            of logs.

5.) **Gearman**       : Optional builtin gearman daemon provided by zuul scheduler.

6.) **Nodepool**      : Provides nodes for Zuul to use whn executing jobs.


Zuul's configuration revolves around the concept of the pipeline. In Zuul, a pipeline encompasses a 
workflow process which can be applied to one or more projects. Pipelines have associated triggers 
which are description of events which should cause something to be enqueued into a pipeline. Once
all of the jobs for an item in a pipeline have been run, the pipeline's reporters are responsible for reporting the results of all the jobs.

The items enqueued into a pipeline are each associated with a git ref. That ref may point to a
proposed change, or it may be the tip of a branch or a tag. The triggerting event determines the
ref, whether it represents a proposed or merged commit.

Zuul has a configuration that tells it what jobs to run for what projects. In typical setup it puts
the jobs it wants to run into gearman for a launcher to consume. gearman is a job-server. Zuul puts
into gearman basically a tuple(job-name, node-type) for each job it wants to run, specifying the 
unique job name run and what type of node it should be run on. A group of Zuul launchers are 
subscribed to gearman as workers. It will consume the job requests from queue and. However, a 
launcher needs two things to be able to run a job - a job definition and a worker node. Job 
definitions stored in external YAML files. The second part is managed by a customized management 
tool called nodepool. Nodepool watches the gearman queue and sees what requests are coming out of
Zuul. It looks at node-type of jobs in the queue and decides what type of nodes need to start and
which cloud providers have capacity to satisfy demand.

The executable contents of jobs themselves are Ansible playbooks. Ansible's support for orchestrating
tasks on remote nodes is particularly used to Zuul's support for multi-node testing. Zuul will wait for
the results of all jobs for the change and post result back to Gerrit/Github, it either gives a 
positive vote or negative vote if jobs failed.
![openstack-ci-zuulv25](https://user-images.githubusercontent.com/13834498/40612458-4223cda6-6298-11e8-88bd-15327cb32af7.png)

**Comment by rohanKanojia** [comment]
+  Deployable on Openshift :x:
     
     Although there is some work going on related to this. Some of the drafts of a spec :
        -  https://review.openstack.org/#/c/560136/
        -  https://review.openstack.org/570667
        -  https://review.openstack.org/535557
     That being said, I'm not sure how long it would take for Zuul to provide production grade support
     for Openshift/Kubernetes deployments.

+ Able to retrieve status of jobs, logs :heavy_check_mark:
         Zuul jobs are implemented as Ansible playbooks. It prepares the repositories used for a job,
         installs any required Ansible roles, and then executes the job's playbooks.

         Before starting each job, the Zuul executor creates a directory to hold all of the content 
         related to the job. The heirarchy is:
             work/            -> The working directory of the job.
             work/src/        -> Contains prepared git repositories for the job
             work/logs/       -> Where the Ansible log for the job is written

+ Orchestration of parallel, serial pipeline steps :heavy_check_mark:
         A particular focus of Zuul is ensuring correctly ordered testing of changes in parallel. A gating
         system should always test change applied to the tip of branch exactly as it is going to be merged
         Zuul's dependent pipeline manager allows for parallel execution of test jobs for gating while
         ensuring changes are tested correctly, exactly as if they had been tested one at a time.

+ Ecosystem of plugin support :x:

+ Canary deployments :x: 

+ Health of upstream :heavy_check_mark:
         Source: https://git.zuul-ci.org/cgit

+ License :heavy_check_mark:
         (Apache License v 2.0)

**Comment by rohanKanojia** [comment]
## Update
I've been able to set up Zuul as a part of [SoftwareFactory](https://softwarefactory-project.io/sf/welcome.html). SF is distribution of RPMs packages with deployment tool called sfconfig that setup a development forge based on Zuul3. Think of relation between Software Factory and Zuul as same as the relation between Openshift and Kubernetes. I setup a demo project using `sfconfig --provision-demo` flags. I installed these following:
 https://softwarefactory-project.io/logs/26/12426/2/check/sf-docs-build/e4a957d/docs-html/contributor/how_to_setup_sf_sandbox.html

I'm trying simple workflows right now and would investigate the proposed draft for Openshift support by Tristan Cacqueray: https://softwarefactory-project.io/draft/zuul-openshift/

**Comment by rohanKanojia** [comment]
So I've been able to successfully setup openshift test job on top of Zuul following https://softwarefactory-project.io/draft/zuul-openshift/.
**This is just a prototype which proves that Openshift builds on top of Zuul is feasible.** 
1.) With Nodepool openshift provider, Nodepool returns an Openshift project with a service account
instead of an ansible inventory instance. 
```
[root@dhcppc1 openshift]# nodepool list
+------------+--------------------+-------------------+--------------------+-------------+------+-------+-------------+----------+
| ID         | Provider           | Label             | Server ID          | Public IPv4 | IPv6 | State | Age         | Locked   |
+------------+--------------------+-------------------+--------------------+-------------+------+-------+-------------+----------+
| 0000000063 | openshift.rdocloud | openshift-project | zuul-ci-0000000063 | None        | None | ready | 00:01:10:15 | unlocked |
+------------+--------------------+-------------------+--------------------+-------------+------+-------+-------------+----------+
[root@dhcppc1 openshift]# 

```

2.) Zuul created a separate namespace for test job, and created all resources in it.
```
[root@dhcppc1 openshift]# oc get projects
NAME                 DISPLAY NAME   STATUS
myproject            My Project     Active
zuul-ci-0000000063                  Active
[root@dhcppc1 openshift]# oc project zuul-ci-0000000063
Already on project "zuul-ci-0000000063" on server "https://127.0.0.1:8443".
[root@dhcppc1 openshift]# oc get pods
NAME                         READY     STATUS      RESTARTS   AGE
demo-project-2-build         1/1       Running     0          29s
zuul-merger-server-1-bpjnb   1/1       Running     0          1m
zuul-merger-server-2-build   0/1       Completed   0          1m
[root@dhcppc1 openshift]# oc logs -f demo-project-2-build
---> Installing application source ...
---> Installing application ...
Processing /opt/app-root/src
Installing collected packages: UNKNOWN
Running setup.py install for UNKNOWN: started
Running setup.py install for UNKNOWN: finished with status 'done'
Successfully installed UNKNOWN-0.0.0
You are using pip version 9.0.1, however version 10.0.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
Pushing image 172.30.1.1:5000/zuul-ci-0000000063/demo-project:latest ...
Pushed 0/10 layers, 30% complete
Pushed 1/10 layers, 29% complete
Pushed 2/10 layers, 30% complete
Pushed 3/10 layers, 30% complete
Pushed 4/10 layers, 40% complete
Pushed 5/10 layers, 50% complete
Pushed 6/10 layers, 63% complete
Pushed 7/10 layers, 74% complete
Pushed 8/10 layers, 94% complete
Pushed 9/10 layers, 98% complete
Pushed 10/10 layers, 100% complete
Push successful

```

3.) Zuul reported status back on Gerrit with the results :
![screenshot_20180613_201020](https://user-images.githubusercontent.com/13834498/41358573-4800eab4-6f46-11e8-9182-6fe09cabe945.png)
Ansible Run Analysis for the job:
![screenshot_20180613_201037](https://user-images.githubusercontent.com/13834498/41358714-9abc19f4-6f46-11e8-988d-fea4c184c2d8.png)
![screenshot_20180613_201058](https://user-images.githubusercontent.com/13834498/41358715-9b256ec2-6f46-11e8-81b8-152ffcfb21a9.png)

**Comment by dmsimard** [comment]
@rohanKanojia (ARA author here) note that ARA can also very easily be implemented with openshift-ansible too -- here's an example CI job for our standalone openshift registry: https://logs.rdoproject.org/81/14181/7/check/legacy-rdo-registry-integration/01ee52c/logs/ara/

**Comment by lordofthejars** [comment]
@sthaha can we close this issue? I guess that for now and in middle term we are going to be stick to Jenkins, so maybe in the future, if we want to change/investigate another thing we could open a new issue or reopen this one.
